flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataStreamProcTimeAggregateGlobalWindowFunction.scala

```diff
@@ -0,0 +1,106 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.aggregate
+
+import org.apache.flink.streaming.api.functions.windowing.RichAllWindowFunction
+import org.apache.flink.types.Row
+import org.apache.flink.util.Collector
+import org.apache.flink.streaming.api.windowing.windows.Window
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.table.functions.Accumulator
+
+import java.lang.Iterable
+import org.apache.flink.table.functions.AggregateFunction
+
+/**
+  * Computes the final aggregate value from incrementally computed aggreagtes.
+  *
+  * @param aggregates The aggregates to be computed
+  * @param aggFields the fields on which to apply the aggregate.
+  * @param forwardedFieldCount The fields to be carried from current row.
+  */
+class DataStreamProcTimeAggregateGlobalWindowFunction[W <: Window](
+  private val aggregates: Array[AggregateFunction[_]],
+  private val aggFields: Array[Int],
+  private val forwardedFieldCount: Int)
+    extends RichAllWindowFunction[Row, Row, W] {
+
+  private var output: Row = _
+  private var accumulators: Row = _
+
+  override def open(parameters: Configuration): Unit = {
+    output = new Row(forwardedFieldCount + aggregates.length)
+    accumulators = new Row(aggregates.length)
+    var i = 0
+    while (i < aggregates.length) {
+      accumulators.setField(i, aggregates(i).createAccumulator())
+      i = i + 1
+    }
+  }
+
+    /**
+    * Calculate aggregated values output by aggregate buffer, and set them into output
+    * Row based on the mapping relation between intermediate aggregate data and output data.
+    */
+  override def apply(
+    window: W,
+    records: Iterable[Row],
+    out: Collector[Row]): Unit = {
+
+   
+    //initialize the values of the aggregators by re-creating them
+    //the design of the Accumulator interface should be extended to enable 
+    //a reset function for better performance
+    var i = 0
+    while (i < aggregates.length) {
+      aggregates(i).resetAccumulator(accumulators.getField(i).asInstanceOf[Accumulator])
+      i += 1
+    }
+    var reuse: Row = null
+    //iterate through the elements and aggregate
+    val iter = records.iterator
+    while (iter.hasNext) {
+      reuse = iter.next
+      i = 0
+      while (i < aggregates.length) {
+        val accumulator = accumulators.getField(i).asInstanceOf[Accumulator]
+        aggregates(i).accumulate(accumulator, reuse.getField(aggFields(i)))
+        i += 1
+      }
+    }
+
+    //set the values of the result with current elements values if needed
```

Comment:
@fhueske I would have said that it is always needed to carry on also the contents of the last event...but this is a too strong statement. Although i did not see examples where that is not the case...there are many SQL examples out there. Also, it might be that the optimizer of calcite changes at some point.